gms/feature_service.cc

```diff
@@ -216,4 +216,26 @@ void feature_service::enable(const std::set<std::string_view>& list) {
     }
 }
 
+struct feature_service::supported_features_change_subscription::impl {
+    feature_service& _service;
+    std::list<features_changed_callback_t>::iterator _it;
+
+    impl(feature_service& service, features_changed_callback_t cb)
+        : _service(service)
+        , _it(_service._supported_features_change_callbacks.insert(_service._supported_features_change_callbacks.begin(), std::move(cb)))
+    {}
+
+    ~impl() {
+        _service._supported_features_change_callbacks.erase(_it);
+    }
+};
```

Comment:
It's not a merge blocker, but there's pretty similar `feature::when_enabled` subscription engine out there. It would be awesome if we don't have the `N+1`th, but generalize the existing (all the more so it's likely to stay inside feature service)

Reply:
Maybe the entire thing deserves a bigger refactor and we could get rid of this `_disabled_features` set, instead storing two booleans in each feature, supported and enabled (or even better, `enum class` with 3 states: `unsupported`, `supported`, `enabled`).

Reply:
No, no, I didn't mean to use per-feature notification. I meant just to generalize the subscription engine and use it for both -- feature::when_enabled() and feature_service::on_supported_features_change() notification

Reply:
The "subscription engine" for `when_enabled` is per-feature: each feature stores a `boost::signals2::signal_type`, and calling `when_enabled` means creating a `boost::signals2::scoped_connection` and a `boost::signals2::signal_type::slot_type` - the slot is connected to the signal and the connection is stored and returned in the subscription object.

(btw. using these complex boost::signals2 concepts here doesn't serve any purpose, I think they could be replaced with a simple list of callbacks like I did here; I'm tempted to refactor this)

On the other hand, the solution for set of supported features is feature-service-wide - we have a single "signal" for the entire service.

If we want a generalized solution to cover both we would probably either have to make the supported notifications on a per-feature basis, or make the enabled notifications feature-service-wide. I think. (Maybe I'm wrong)

Now, gossiper wants to observe all changes to the set of supported features, hence I made the supported thing feature-service-wide.

Reply:
+1 for many of the suggestions here (instead getting rid of boost::signals2).